src/main/scala/org/fortysevendeg/sparksftp/ReadingSFTPConnectorApp.scala

```diff
@@ -18,41 +17,66 @@ object ReadingSFTPConnectorApp extends IOApp {
     ConfigLoader[IO]
       .loadConfig[ReadingSFTPConfig]
 
-  def run(args: List[String]): IO[ExitCode] =
-    for {
-      config: ReadingSFTPConfig <- setupConfig
-      defaultSparkConf: SparkConf = SparkUtils.createSparkConfWithSFTPSupport(config)
-      sparkSession: SparkSession = SparkSession.builder
+  def run(args: List[String]): IO[ExitCode] = {
+
+    def createSparkSession(config: ReadingSFTPConfig): IO[SparkSession] = IO {
+      val defaultSparkConf: SparkConf = createSparkConfWithSFTPSupport(config)
+      SparkSession.builder
         .config(defaultSparkConf)
         .enableHiveSupport
         .getOrCreate()
+    }
 
-      sftpConfig = configs.SFTPConfig
-        .configFromContextProperties(sparkSession.sparkContext, config.sftp)
-
-      // Read the source files from SFTP into dataframes
-      users = dataframeFromCsvWithSFTPConnector(sparkSession, sftpConfig, sftpConfig.sftpUserPath)
-      salaries = dataframeFromCsvWithSFTPConnector(
+    def readDataFramesWithSFTPConnector(
+        sparkSession: SparkSession,
+        sftpConfig: SFTPConfig
+    ): IO[(DataFrame, DataFrame)] = {
+      val users =
+        dataframeFromCsvWithSFTPConnector(sparkSession, sftpConfig, sftpConfig.sftpUserPath)
```

Comment:
Seems like the `dataframeFromCsvWithSFTPConnector` is where the _effect_ is happening. I'd suggest modifying that function (instead of doing a final `IO((a, b))`

Reply:
Changed in: a78818e